[Zi Hung] iP

zihung20 opened the pull request:
# Lolok  
A chatbot used for recording daily life tasks  

## Features  

1. **Add** tasks to your data. There are three types of tasks: **Todo**, **Deadline**, and **Event**.  
2. **Mark** or **unmark** tasks as done/undone.  
3. **Automatic saving** ensures tasks persist between sessions—**no need** to re-enter them!  

> *"Stay organized today, so your future self thanks you tomorrow."* 🚀  

## Software Engineering Principles  

Lolok follows strong **software engineering** and **OOP principles** to ensure maintainability and scalability:  

- **Encapsulation** – Keeps data secure and prevents unintended modifications.  
- **Abstraction** – Simplifies task management logic.  
- **Inheritance** – Enhances reusability for different task types.  
- **Polymorphism** – Allows flexible behavior across various task operations.  
- **SOLID Principles** – Ensures clean architecture and modular design.  

## Tech Stack  

Lolok is built using modern tools for efficiency and maintainability:  

- **`Java`** – Core programming language.  
- **`JavaFX`** + **`SceneBuilder`** – For an interactive UI.  
- **`Gradle`** – Build automation and dependency management.  
- **`JUnit`**  – Unit testing framework.  

## Sample code
```java
/**
 * Prints the tasks in the list.
 */
public void printList() {
    StringBuilder builder = new StringBuilder("Here are the tasks in your list: \n");
    for (int i = 0; i < list.size(); i++) {
        builder.append(String.format("%d. %s \n", (i + 1), list.get(i).toString()));
    }
    Ui.printMessage(builder.toString());
}
```

## Future plan
- [ ] Multi-Platform Integration
- [x] Advanced Natural Language Understanding
- [ ] convince [Google](https://www.google.com/) to integrate our chatbot into their system

---

Lolok is designed to be **robust, user-friendly, and efficient**, making task management effortless!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

zihung20 committed:
Initialize chatbot name and add level 0 features

zihung20 committed:
Add echo features

zihung20 committed:
Add ability to store user text

zihung20 committed:
Create Task class to represent Task

zihung20 committed:
Add mark task feature

zihung20 committed:
Create class file for todo, deadline and event

zihung20 committed:
Create polymorphism for task display

zihung20 committed:
Add textUiTesting test case

zihung20 committed:
Add delete task functions

zihung20 committed:
Restructure code to facilitate implementation of Exception and Deletion

zihung20 committed:
Update test case for deletion

zihung20 committed:
Update Exception handling for deletion and invalid command

zihung20 committed:
Created file for ENUM feature

zihung20 committed:
Created file for storing data

zihung20 committed:
Add load data features at the start of executing file

zihung20 committed:
Add polymorphism to Task class when creating formatted string

zihung20 committed:
Implement save tasks to file functionality

zihung20 committed:
Add more detailed error message

zihung20 committed:
Save date with java LocaleDateTime class

zihung20 committed:
Handle exceptions when parsing date strings

zihung20 committed:
Merge branch 'branch-Level-8'

zihung20 committed:
Create Storage class for better OOP structure

zihung20 committed:
Create TaskList class for better OOP structure

zihung20 committed:
Create Ui class for better OOP structure

zihung20 committed:
Move command related method to Command class

zihung20 committed:
Fix Ui problem

zihung20 committed:
Delete all unuse imports

zihung20 committed:
Add all files to a package

zihung20 committed:
Fix bug when loading data

zihung20 committed:
Merge branch 'add-gradle-support'

zihung20 committed:
Use gradle to automate build process

zihung20 committed:
Add Junit test cases

zihung20 committed:
Change name of jar file

zihung20 committed:
Add java doc to Command class

zihung20 committed:
Add java doc to Task and Deadline class

zihung20 committed:
Add java docs to Event and Task class

zihung20 committed:
Add java docs to the Lolol class and Storage class

zihung20 committed:
Add java docs to TaskList and Todo

zihung20 committed:
Format code according to code style

zihung20 committed:
Add search method by keyword for TaskList

zihung20 committed:
Add feature to find tasks by keyword

zihung20 committed:
Backup

zihung20 committed:
Merge branch 'branch-A-JavaDoc'

zihung20 committed:
Merge branch 'branch-A-CodingStandard'

zihung20 committed:
Merge branch 'branch-Level-9'

iuhiah commented on src/main/java/lolok/Command.java:
```diff
@@ -0,0 +1,112 @@
+
+    private String getType() {
+        return this.type;
+    }
```
```suggestion
```

Looks like a redundant private getter here, it doesn't perform any meaningful actions beside returning the raw attribute. Consider just referring to the attribute directly.

iuhiah commented on src/main/java/lolok/Command.java:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+
+    private void multipleArgumentCommand(TaskList taskList) {
```
I think this sounds more like a class name since it's a noun, perhaps something like executeMultipleArgumentCommand() would be better?

loyaltypollution replied:
This allows us to classify commands together! That's something I didn't think of 👍

Perhaps we could consider turning this into a class in the future?
DeleteCommand - extends -> MultipleArgumentCommand - extends -> Command

zihung20 committed:
Add configuration file for checkstyle

zihung20 committed:
Make changes to adhere to the coding style

zihung20 committed:
Merge branch 'A-CheckStyle'

zihung20 committed:
Restructure the task related class into a package

zihung20 committed:
Add exception to exception package and TaskList class to task package

zihung20 committed:
Add Action and Command to command package

zihung20 committed:
Merge branch 'fix-structure'

zihung20 committed:
Merge branch 'master' into Level-10

zihung20 committed:
Add basic files for GUI

zihung20 committed:
Set up a basic window for the chatbot, but no logic implemented

zihung20 committed:
Change Task to an abstract class

zihung20 committed:
Add basic GUI for Lolok

loyaltypollution commented on src/main/java/lolok/Lolok.java:
```diff
@@ -0,0 +1,67 @@
+
+    private void exit() {
+        this.storage.saveData(taskList.getList(), false);
+        System.out.println("Bye. Hope to see you again soon!");
```
Could this line be handled by Ui instead? 

This abstracts away System.out for ease of other forms of ui implementation in future.

loyaltypollution commented on src/main/java/lolok/Lolok.java:
```diff
@@ -0,0 +1,67 @@
+                command.executeCommand(taskList, ui, storage);
+                exit = command.isExit();
+            } catch (LolokException e) {
+                // should already handle exception in duke.Command
```
Perhaps the comment refers to lolok.Command?

loyaltypollution commented on src/main/java/lolok/Storage.java:
```diff
@@ -0,0 +1,53 @@
+    public void saveData(List<Task> tasks, boolean append) {
+        System.out.println("Saving tasks...");
+        //FileWriter create file when it does not exist
+        try (FileWriter fw = new FileWriter(this.filePath, append)) {
```
Perhaps we can have filepath instead?
```suggestion
        try (FileWriter fw = new FileWriter(filePath, append)) {
```
Our convention guide describes usage of this [here](https://se-education.org/guides/conventions/java/index.html#variables)

zihung20 replied:
ohhh I forgot that rules, thanks for pointing out

loyaltypollution commented on src/main/java/lolok/Command.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    public Command(String[] block) {
+        this.blocks = block;
+        this.type = block[0];
```
Effective method of parsing Commands! Good job here

loyaltypollution reviewed (commented):
LGTM. Just a few nits to fix. Great work

iuhiah reviewed (commented):
Code is overall well abstracted but would recommend further refactoring the project structure to split files into relevant packages (e.g. task/exceptions)!

zihung20 committed:
Fix the implementation of printing messages

zihung20 committed:
Add font style and style dialog box

zihung20 committed:
Add styling to chatbot

zihung20 committed:
Merge branch 'Level-10'

zihung20 committed:
Use varargs to Ui class

zihung20 committed:
Fix excessive use of "this" in task package

zihung20 committed:
Fix excessive use of "this" in other package

zihung20 committed:
Merge branch 'A-Varargs'

zihung20 committed:
Fix merging issue with branch

zihung20 committed:
Refactor task list methods to use streams

zihung20 committed:
Add assertion to ensure get command arguments are used correctly

When retrieving arguments, this function should only be used for
commands that require arguments.

To prevent misuse, an assertion has been added to ensure the
argument count is greater than zero.

assert the count must be > 0 when using this function

This is the simplest way to enforce correct usage and prevent
errors caused by improper function calls.

zihung20 committed:
Refactor argument handling to avoid magic numbers and improve safety

Previously, the argument count was manually set and retrieved using
magic numbers.

To improve code quality and reduce the potential for misuse, the
argument count is now handled by the Action Enum.

This refactor ensures safer access to the required number of
arguments, making the code more maintainable and reducing human error.

zihung20 committed:
Change datetime format string as a constant

zihung20 committed:
Remove unused variable and redundant initialization in Lolok class

zihung20 committed:
Merge pull request #1 from zihung20/A-Assertions

Add assertion to ensure get command arguments are used correctly

zihung20 committed:
Merge branch 'master' of https://github.com/zihung20/ip

zihung20 committed:
Merge branch 'master' into A-Streams

zihung20 committed:
Merge branch 'master' into A-CodeQuality

# Conflicts:
#	src/main/java/lolok/command/Command.java

zihung20 committed:
Merge pull request #2 from zihung20/A-CodeQuality

A code quality

zihung20 committed:
Merge branch 'master' into A-Streams

zihung20 committed:
Merge pull request #3 from zihung20/A-Streams

zihung20 committed:
Create yml file for github CI

zihung20 committed:
Prevent duplicate tasks from being added

zihung20 committed:
Fix inaccurate command interpretation

zihung20 committed:
Align test file structure with Java source file structure

zihung20 committed:
Suppress Javadoc style checking

zihung20 committed:
Add greeting message when start the app

zihung20 committed:
Remove testing runtest.bat

zihung20 committed:
Merge branch 'master' into branch-A-CheckStyle

zihung20 committed:
Fix suppression of header comment error for Javadoc

zihung20 committed:
Merge branch 'branch-A-CheckStyle'

zihung20 committed:
Make a better GUI for user and chatbot dialog

zihung20 committed:
Improve dialog box UI

zihung20 committed:
Merge pull request #5 from zihung20/branch-A-BetterGui

Branch a better gui

zihung20 committed:
Add citation of using Ai tools

zihung20 committed:
Set title and icon for the JavaFX window

zihung20 committed:
Add representative screenshot of the chatbot

zihung20 committed:
Add user guide for Lolok chatbot in GFMD format

zihung20 committed:
Merge branch 'branch-A-UserGuide'

zihung20 committed:
Fix incorrect example position

zihung20 committed:
Merge branch 'master' into branch-A-CheckStyle

zihung20 committed:
Fix style violations

zihung20 committed:
Merge branch 'branch-A-CheckStyle'

zihung20 committed:
Fix data file location issue in production environment

zihung20 committed:
Add some test case for command

zihung20 committed:
Refactor code for readability and improve error messaging

zihung20 committed:
Improve exception handling and refine code using AI

zihung20 committed:
Add more test cases for better coverage of task package

zihung20 committed:
Merge pull request #6 from zihung20/branch-A-AiAssisted

Branch a ai assisted

zihung20 committed:
Add find feature on user guide

zihung20 committed:
Fix format for find command output

zihung20 committed:
Add secret personality to chatbot

zihung20 committed:
Add abstraction layer when instantiate chatbot dialog

zihung20 committed:
Finalize chatbot name

zihung20 committed:
Update showcase picture

zihung20 committed:
Fix chatbot name in user guide
